File: app/services/savings_service.py
import math

def calculate_percentage_saving(amount: float, ratio: float) -> float:
    """
    Harcama tutarının belirli bir yüzdesini hesaplayarak birikim tutarını döner.
    Örn: 49.99 TL harcama ve %10 oran -> 5.00 TL birikim (yuvarlanmış)
    
    :param amount: Harcama tutarı
    :param ratio: Birikim oranı (0-100 arası, örn: 10)
    :return: Birikim tutarı
    """
    if not amount or amount <= 0 or not ratio or ratio <= 0:
        return 0.0

    raw_saving = amount * (ratio / 100)
    
    # Python'da finansal yuvarlama için f-string formatlama en güvenli yoldur.
    # toFixed(2) karşılığıdır.
    return float(f"{raw_saving:.2f}")

def calculate_round_up(amount: float, step: float) -> float:
    """
    Klasik Yuvarlama Mantığı (Round-up)
    Tutarın bir sonraki 'step' katına olan uzaklığını hesaplar.
    Örn: 42.30 TL harcama ve step: 10 -> 50'ye tamamlar, 7.70 TL döner.
    
    :param amount: Harcama tutarı
    :param step: Yuvarlama hedefi (10, 50, 100 vb.)
    :return: Birikim tutarı
    """
    if not amount or amount <= 0 or not step or step <= 0:
        return 0.0

    target = math.ceil(amount / step) * step
    savings = target - amount
    
    return float(f"{savings:.2f}")

File: src/services/savingsService.js
/**
 * Kurusla - Savings Service (Mikro Birikim Motoru)
 * Bu servis, kullanıcının harcamalarından yapılacak birikim tutarlarını hesaplar.
 */

/**
 * Harcama tutarının belirli bir yüzdesini hesaplayarak birikim tutarını döner.
 * @param {number} amount - Harcama tutarı
 * @param {number} ratio - Birikim oranı (0-100 arası, örn: 10)
 * @returns {number} - Birikim tutarı
 */
const calculatePercentageSaving = (amount, ratio) => {
  if (!amount || amount <= 0 || !ratio || ratio <= 0) return 0;

  const rawSaving = amount * (ratio / 100);
  return parseFloat(rawSaving.toFixed(2));
};

/**
 * Klasik Yuvarlama Mantığı (Round-up)
 * @param {number} amount - Harcama tutarı
 * @param {number} step - Yuvarlama hedefi (10, 50, 100 vb.)
 * @returns {number} - Birikim tutarı
 */
const calculateRoundUp = (amount, step) => {
  if (!amount || amount <= 0 || !step || step <= 0) return 0;

  const target = Math.ceil(amount / step) * step;
  const savings = target - amount;
  return parseFloat(savings.toFixed(2));
};

module.exports = {
  calculatePercentageSaving,
  calculateRoundUp
};
